File: src/utils/assistantHealth.ts
export type HealthStatus = 'healthy' | 'degraded' | 'unhealthy' | 'offline'
export type HealthCheckType = 'provider' | 'session' | 'circuit' | 'heartbeat'

export interface HealthCheckResult {
  type: HealthCheckType
  status: HealthStatus
  message: string
  timestamp: number
  latencyMs?: number
  details?: Record<string, unknown>
}

export interface AssistantHealthSnapshot {
  overall: HealthStatus
  checks: HealthCheckResult[]
  lastCheckedAt: number
  consecutiveFailures: number
  lastHealthyAt: number | null
  lastUnhealthyAt: number | null
  uptimeMs: number
  totalRequests: number
  totalFailures: number
  totalSuccesses: number
}

type HealthChangeListener = (snapshot: AssistantHealthSnapshot) => void

const listeners = new Set<HealthChangeListener>()
let snapshot: AssistantHealthSnapshot = createInitialSnapshot()
let startedAt = Date.now()

function createInitialSnapshot(): AssistantHealthSnapshot {
  return {
    overall: 'offline',
    checks: [],
    lastCheckedAt: 0,
    consecutiveFailures: 0,
    lastHealthyAt: null,
    lastUnhealthyAt: null,
    uptimeMs: 0,
    totalRequests: 0,
    totalFailures: 0,
    totalSuccesses: 0,
  }
}

function computeOverall(checks: HealthCheckResult[]): HealthStatus {
  if (checks.length === 0) return 'offline'
  const statuses = checks.map((c) => c.status)
  if (statuses.some((s) => s === 'unhealthy' || s === 'offline')) return 'unhealthy'
  if (statuses.some((s) => s === 'degraded')) return 'degraded'
  return 'healthy'
}

function notifyListeners(): void {
  for (const listener of listeners) {
    try {
      listener(snapshot)
    } catch {}
  }
}

export const assistantHealth = {
  start(): void {
    startedAt = Date.now()
    snapshot = createInitialSnapshot()
    snapshot.overall = 'healthy'
    snapshot.lastHealthyAt = Date.now()
  },

  recordCheck(result: HealthCheckResult): void {
    snapshot.checks = snapshot.checks.filter((c) => c.type !== result.type)
    snapshot.checks.push(result)
    snapshot.lastCheckedAt = Date.now()
    snapshot.overall = computeOverall(snapshot.checks)

    if (result.status === 'healthy') {
      snapshot.consecutiveFailures = 0
      snapshot.lastHealthyAt = Date.now()
    } else {
      snapshot.consecutiveFailures++
      snapshot.lastUnhealthyAt = Date.now()
    }

    snapshot.uptimeMs = Date.now() - startedAt
    notifyListeners()
  },

  recordRequest(success: boolean): void {
    snapshot.totalRequests++
    if (success) {
      snapshot.totalSuccesses++
    } else {
      snapshot.totalFailures++
    }
    snapshot.uptimeMs = Date.now() - startedAt
  },

  getSnapshot(): AssistantHealthSnapshot {
    return { ...snapshot, checks: [...snapshot.checks] }
  },

  getStatus(): HealthStatus {
    return snapshot.overall
  },

  isHealthy(): boolean {
    return snapshot.overall === 'healthy' || snapshot.overall === 'degraded'
  },

  shouldRecover(): boolean {
    return snapshot.consecutiveFailures >= 3 || snapshot.overall === 'unhealthy'
  },

  reset(): void {
    snapshot = createInitialSnapshot()
    snapshot.overall = 'healthy'
    snapshot.lastHealthyAt = Date.now()
    notifyListeners()
  },

  onHealthChange(listener: HealthChangeListener): () => void {
    listeners.add(listener)
    return () => listeners.delete(listener)
  },

  createProviderCheck(provider: string, latencyMs: number, error?: string): HealthCheckResult {
    let status: HealthStatus = 'healthy'
    let message = `Provider ${provider} responding normally`
    if (error) {
      status = latencyMs > 10000 ? 'unhealthy' : 'degraded'
      message = `Provider ${provider} error: ${error}`
    } else if (latencyMs > 10000) {
      status = 'degraded'
      message = `Provider ${provider} slow response (${latencyMs}ms)`
    }
    return {
      type: 'provider',
      status,
      message,
      timestamp: Date.now(),
      latencyMs,
      details: { provider, error },
    }
  },

  createCircuitCheck(circuitState: string, tripCount: number): HealthCheckResult {
    let status: HealthStatus = 'healthy'
    let message = `Circuit breaker closed`
    if (circuitState === 'open') {
      status = 'unhealthy'
      message = `Circuit breaker OPEN (${tripCount} trips)`
    } else if (circuitState === 'half_open') {
      status = 'degraded'
      message = `Circuit breaker HALF-OPEN (probing)`
    }
    return {
      type: 'circuit',
      status,
      message,
      timestamp: Date.now(),
      details: { circuitState, tripCount },
    }
  },

  createSessionCheck(activeSessions: number, stalledSessions: number): HealthCheckResult {
    let status: HealthStatus = 'healthy'
    let message = `${activeSessions} active sessions`
    if (stalledSessions > 0) {
      status = stalledSessions > 2 ? 'unhealthy' : 'degraded'
      message = `${stalledSessions} stalled of ${activeSessions} sessions`
    }
    return {
      type: 'session',
      status,
      message,
      timestamp: Date.now(),
      details: { activeSessions, stalledSessions },
    }
  },

  createHeartbeatCheck(responsive: boolean, latencyMs?: number): HealthCheckResult {
    const status: HealthStatus = responsive ? 'healthy' : 'unhealthy'
    const message = responsive
      ? `Assistant responsive${latencyMs ? ` (${latencyMs}ms)` : ''}`
      : 'Assistant not responding'
    return { type: 'heartbeat', status, message, timestamp: Date.now(), latencyMs }
  },
}

File: frontend/src/utils/assistant_health.rs
//! Assistant health monitoring — ported from src/utils/assistantHealth.ts
//!
//! Tracks health of the LLM assistant via provider checks, session checks,
//! circuit breaker checks, and heartbeat checks.

use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HealthStatus {
    Healthy,
    Degraded,
    Unhealthy,
    Offline,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HealthCheckType {
    Provider,
    Session,
    Circuit,
    Heartbeat,
}

#[derive(Debug, Clone)]
pub struct HealthCheckResult {
    pub check_type: HealthCheckType,
    pub status: HealthStatus,
    pub message: String,
    pub timestamp: u64,
    pub latency_ms: Option<u64>,
    pub details: Option<serde_json::Value>,
}

#[derive(Debug, Clone)]
pub struct AssistantHealthSnapshot {
    pub overall: HealthStatus,
    pub checks: Vec<HealthCheckResult>,
    pub last_checked_at: u64,
    pub consecutive_failures: u32,
    pub last_healthy_at: Option<u64>,
    pub last_unhealthy_at: Option<u64>,
    pub uptime_ms: u64,
    pub total_requests: u64,
    pub total_failures: u64,
    pub total_successes: u64,
}

pub type HealthChangeListener = Arc<dyn Fn(&AssistantHealthSnapshot) + Send + Sync>;

// ---------------------------------------------------------------------------
// Internal
// ---------------------------------------------------------------------------

struct Inner {
    snapshot: AssistantHealthSnapshot,
    started_at: Option<u64>,
    listeners: Vec<HealthChangeListener>,
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn compute_overall(checks: &[HealthCheckResult]) -> HealthStatus {
    if checks.is_empty() {
        return HealthStatus::Offline;
    }
    let mut has_degraded = false;
    for c in checks {
        match c.status {
            HealthStatus::Unhealthy | HealthStatus::Offline => return HealthStatus::Unhealthy,
            HealthStatus::Degraded => has_degraded = true,
            HealthStatus::Healthy => {}
        }
    }
    if has_degraded {
        HealthStatus::Degraded
    } else {
        HealthStatus::Healthy
    }
}

fn initial_snapshot() -> AssistantHealthSnapshot {
    AssistantHealthSnapshot {
        overall: HealthStatus::Offline,
        checks: Vec::new(),
        last_checked_at: 0,
        consecutive_failures: 0,
        last_healthy_at: None,
        last_unhealthy_at: None,
        uptime_ms: 0,
        total_requests: 0,
        total_failures: 0,
        total_successes: 0,
    }
}

// ---------------------------------------------------------------------------
// AssistantHealth
// ---------------------------------------------------------------------------

pub struct AssistantHealth {
    inner: Arc<Mutex<Inner>>,
}

impl Default for AssistantHealth {
    fn default() -> Self {
        Self::new()
    }
}

impl AssistantHealth {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(Mutex::new(Inner {
                snapshot: initial_snapshot(),
                started_at: None,
                listeners: Vec::new(),
            })),
        }
    }

    pub fn start(&self) {
        let (snapshot, listeners) = {
            let mut inner = self.inner.lock().unwrap();
            let now = now_ms();
            inner.snapshot = initial_snapshot();
            inner.snapshot.overall = HealthStatus::Healthy;
            inner.snapshot.last_healthy_at = Some(now);
            inner.started_at = Some(now);
            (inner.snapshot.clone(), inner.listeners.clone())
        };
        self.notify(snapshot, listeners);
    }

    pub fn record_check(&self, result: HealthCheckResult) {
        let (snapshot, listeners) = {
            let mut inner = self.inner.lock().unwrap();
            let ts = result.timestamp;

            // Replace existing check of the same type
            inner
                .snapshot
                .checks
                .retain(|c| c.check_type != result.check_type);
            inner.snapshot.checks.push(result);
            inner.snapshot.last_checked_at = ts;
            inner.snapshot.overall = compute_overall(&inner.snapshot.checks);

            match inner.snapshot.overall {
                HealthStatus::Healthy | HealthStatus::Degraded => {
                    inner.snapshot.last_healthy_at = Some(ts);
                    inner.snapshot.consecutive_failures = 0;
                }
                HealthStatus::Unhealthy | HealthStatus::Offline => {
                    inner.snapshot.last_unhealthy_at = Some(ts);
                    inner.snapshot.consecutive_failures += 1;
                }
            }

            if let Some(started) = inner.started_at {
                inner.snapshot.uptime_ms = ts.saturating_sub(started);
            }
            (inner.snapshot.clone(), inner.listeners.clone())
        };
        self.notify(snapshot, listeners);
    }

    pub fn record_request(&self, success: bool) {
        let mut inner = self.inner.lock().unwrap();
        inner.snapshot.total_requests += 1;
        if success {
            inner.snapshot.total_successes += 1;
        } else {
            inner.snapshot.total_failures += 1;
        }
    }

    pub fn get_snapshot(&self) -> AssistantHealthSnapshot {
        let inner = self.inner.lock().unwrap();
        inner.snapshot.clone()
    }

    pub fn get_status(&self) -> HealthStatus {
        let inner = self.inner.lock().unwrap();
        inner.snapshot.overall
    }

    pub fn is_healthy(&self) -> bool {
        matches!(
            self.get_status(),
            HealthStatus::Healthy | HealthStatus::Degraded
        )
    }

    pub fn should_recover(&self) -> bool {
        let inner = self.inner.lock().unwrap();
        inner.snapshot.consecutive_failures >= 3
            || inner.snapshot.overall == HealthStatus::Unhealthy
    }

    pub fn reset(&self) {
        let (snapshot, listeners) = {
            let mut inner = self.inner.lock().unwrap();
            inner.snapshot = initial_snapshot();
            inner.started_at = None;
            (inner.snapshot.clone(), inner.listeners.clone())
        };
        self.notify(snapshot, listeners);
        // Re-start immediately
        self.start();
    }

    pub fn on_health_change(&self, listener: HealthChangeListener) {
        let mut inner = self.inner.lock().unwrap();
        inner.listeners.push(listener);
    }

    fn notify(&self, snapshot: AssistantHealthSnapshot, listeners: Vec<HealthChangeListener>) {
        for listener in &listeners {
            listener(&snapshot);
        }
    }
}

// ---------------------------------------------------------------------------
// Factory helpers
// ---------------------------------------------------------------------------

impl AssistantHealth {
    pub fn create_provider_check(
        provider: &str,
        latency_ms: u64,
        error: Option<&str>,
    ) -> HealthCheckResult {
        let (status, message) = match error {
            Some(err) if latency_ms > 10_000 => (
                HealthStatus::Unhealthy,
                format!("Provider {} error with high latency: {}", provider, err),
            ),
            Some(err) => (
                HealthStatus::Degraded,
                format!("Provider {} error: {}", provider, err),
            ),
            None if latency_ms > 10_000 => (
                HealthStatus::Degraded,
                format!("Provider {} slow response ({}ms)", provider, latency_ms),
            ),
            None => (
                HealthStatus::Healthy,
                format!("Provider {} responding normally", provider),
            ),
        };
        HealthCheckResult {
            check_type: HealthCheckType::Provider,
            status,
            message,
            timestamp: now_ms(),
            latency_ms: Some(latency_ms),
            details: None,
        }
    }

    pub fn create_circuit_check(circuit_state: &str, trip_count: u32) -> HealthCheckResult {
        let (status, message) = match circuit_state {
            "open" => (
                HealthStatus::Unhealthy,
                format!("Circuit breaker open ({} trips)", trip_count),
            ),
            "half_open" => (
                HealthStatus::Degraded,
                format!("Circuit breaker half-open ({} trips)", trip_count),
            ),
            _ => (HealthStatus::Healthy, "Circuit breaker closed".to_string()),
        };
        HealthCheckResult {
            check_type: HealthCheckType::Circuit,
            status,
            message,
            timestamp: now_ms(),
            latency_ms: None,
            details: Some(serde_json::json!({ "tripCount": trip_count })),
        }
    }

    pub fn create_session_check(active: u32, stalled: u32) -> HealthCheckResult {
        let (status, message) = if stalled == 0 {
            (
                HealthStatus::Healthy,
                format!("{} active sessions, no stalls", active),
            )
        } else if stalled <= 2 {
            (
                HealthStatus::Degraded,
                format!("{} active, {} stalled", active, stalled),
            )
        } else {
            (
                HealthStatus::Unhealthy,
                format!("{} active, {} stalled", active, stalled),
            )
        };
        HealthCheckResult {
            check_type: HealthCheckType::Session,
            status,
            message,
            timestamp: now_ms(),
            latency_ms: None,
            details: Some(serde_json::json!({ "active": active, "stalled": stalled })),
        }
    }

    pub fn create_heartbeat_check(responsive: bool, latency_ms: Option<u64>) -> HealthCheckResult {
        let (status, message) = if responsive {
            (HealthStatus::Healthy, "Heartbeat responsive".to_string())
        } else {
            (
                HealthStatus::Unhealthy,
                "Heartbeat not responsive".to_string(),
            )
        };
        HealthCheckResult {
            check_type: HealthCheckType::Heartbeat,
            status,
            message,
            timestamp: now_ms(),
            latency_ms,
            details: None,
        }
    }
}
